include/astshim/FrameDict.h

```diff
@@ -0,0 +1,310 @@
+
+@warning FrameDict.getClassName returns "FrameSet". This is because FrameDict has no direct AST equivalent;
+it is merely a convenience wrapper around FrameSet.
+
+@note
+- AST casts all Frame domains to uppercase. This is one reason why domain comparison
+and domain lookup are case blind.
+- Some AST frame classes have default domain names, e.g. SkyFrame defaults to "SKY". Such default
+names are ignored in order to reduce the chance of accidental collisions.
+- This class is named "FrameDict" instead of "FrameMap", even though C++ lookup tables are usually
+std::map or unordered_map, in order to avoid confusion with AST Mapping.
+
+### Attributes
+
+All those of FrameSet.
+*/
+class FrameDict : public FrameSet {
+    friend class Object;
+
+public:
+    /**
+    Construct a FrameDict from a single Frame
+
+    The frame is deep copied.
+
+    @param[in] frame  the first @ref Frame to be inserted into the @ref FrameDict.
+                    This initially becomes both the base and the current Frame.
+                    Further Frames may be added using @ref addFrame
+    @param[in] options  Comma-separated list of attribute assignments.
+    */
+    explicit FrameDict(Frame const &frame, std::string const &options = "") : FrameSet(frame, options) {
+        _rebuildDict();
+    }
+
+    /**
+    Construct a FrameDict from two frames and a mapping that connects them
+
+    Both frames and the mapping are deep copied.
+
+    @param[in] baseFrame  base @ref Frame.
+    @param[in] mapping  mapping connecting baseFrame to currentFrame.
+    @param[in] currentFrame  current @ref Frame.
+    @param[in] options  Comma-separated list of attribute assignments.
+
+    @throws std::invalid_argument if both Frames have the same non-empty domain.
+    */
+    explicit FrameDict(Frame const &baseFrame, Mapping const &mapping, Frame const &currentFrame,
+                       std::string const &options = "")
+            : FrameSet(baseFrame, mapping, currentFrame) {
+        _rebuildDict();
+    }
+
+    /**
+    Construct a FrameDict from a FrameSet
+
+    The FrameSet is deep-copied
+
+    @throws std::invalid_argument if two Frames in the FrameSet have the same non-empty domain.
+    */
+    explicit FrameDict(FrameSet const &frameSet)
+            : FrameDict(reinterpret_cast<AstFrameSet *>(frameSet.copy()->getRawPtr())) {}
+    virtual ~FrameDict() {}
+
+    FrameDict(FrameDict const &) = delete;
+    FrameDict(FrameDict &&) = default;
+    FrameDict &operator=(FrameDict const &) = delete;
+    FrameDict &operator=(FrameDict &&) = default;
+
+    /// Return a deep copy of this object.
+    std::shared_ptr<FrameSet> copy() const { return std::static_pointer_cast<FrameDict>(copyPolymorphic()); }
+
+    void addFrame(int iframe, Mapping const &map, Frame const &frame) override {
+        FrameSet::addFrame(iframe, map, frame);
+        _addFrameToDict(frame, FrameSet::getCurrent());
```

Comment:
Good catch. I propose the following fix (for this and other mutating methods that need more safety):
- Test that the new frame is safe to add to the dict before calling `FrameSet::addFrame`
- Add a `doAssert` argument to `_addFrameToDict` and `_rebuildDict`. If true abort on a name conflict; this indicates that the logic is buggy and the internal state is corrupted. True is used for all methods. False is used for constructors, so that they raise an exception with a helpful message if the provided data isn't usable.